[stable-2.8] User - Create parent directories if they do not exist in the specified home path (#51043)

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Backport of #51043 for Ansible 2.8
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/modules/system/user.py`

samdoran committed:
[stable-2.8] User - Create parent directories if they do not exist in the specified home path (#51043)

* Create a user home directory if it has parents that do not exist

The useradd command line tool does not create parent directories. Check if the specified home path has parents that do not exist. If so, create them prior to running useradd, then set the proper permission on the created directory.

Add tests

Signed-off-by: Sam Doran <sdoran@redhat.com>

* Use dict for default user group in tests

Signed-off-by: Sam Doran <sdoran@redhat.com>

* Fix tests

Signed-off-by: Sam Doran <sdoran@redhat.com>
(cherry picked from commit d2edf1d435)

Co-authored-by: Sam Doran <sdoran@redhat.com>

abadger commented:
Mentioned this on slack... before this gets merged, it needs to be examined for potential security issues.  Anytime we create parent directories we open the door to malicious users on otherwise good machines being able to cause mischief (for instance, via symlink attacks in world writable dirs).  Have to carefully audit the code for potential ways to abuse it before enabling it.

mgedmin committed:
user: createhome=no home=/no/such/dir (#60310)

Fixes #60307.

This bug was introduced in commit d2edf1d43531f938ccbdef79c927141ba193b96d
("User - Create parent directories if they do not exist in the specified
home path (#51043)") and did not make it into any releases.

(cherry picked from commit c71622b31a52c44f171a4bb9da1331939ad7aa60)

abadger commented:
@samdoran When you next have some time, let's go over the possible security ramifications on slack and decide whether to document those cases or whether all of these cases in the user module should be failures instead.

samdoran commented:
@abadger Sounds like a plan. I did some testing based on the symlink attack you described and I'd like to go over the results with you.
